### PR TITLE
Use backwards incompatible change from ansible-compat 4.0.1

### LINKED ISSUE #3903: `molecule` broken with latest version of `ansible-compat`
# Issue Type

- Bug report

# Molecule and Ansible details

```console
$ ansible --version && molecule --version
ansible [core 2.13.9]
  config file = None
  configured module search path = ['/home/mcdonnnj/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/ansible
  ansible collection location = /home/mcdonnnj/.ansible/collections:/usr/share/ansible/collections
  executable location = /home/mcdonnnj/.pyenv/versions/ansible-role-docker/bin/ansible
  python version = 3.11.3 (main, Apr 10 2023, 14:31:58) [GCC 12.2.0]
  jinja version = 3.1.2
  libyaml = True
molecule 5.0.0 using python 3.11
    ansible:2.13.9
    azure:23.4.1 from molecule_plugins
    containers:23.4.1 from molecule_plugins requiring collections: ansible.posix>=1.3.0 community.docker>=1.9.1 containers.podman>=1.8.1
    delegated:5.0.0 from molecule
    docker:23.4.1 from molecule_plugins requiring collections: community.docker>=3.0.2 ansible.posix>=1.4.0
    ec2:23.4.1 from molecule_plugins
    gce:23.4.1 from molecule_plugins requiring collections: google.cloud>=1.0.2 community.crypto>=1.8.0
    podman:23.4.1 from molecule_plugins requiring collections: containers.podman>=1.7.0 ansible.posix>=1.3.0
    vagrant:23.4.1 from molecule_plugins
```

Molecule installation method (one of):

- pip

Ansible installation method (one of):

- pip

Detail any linters or test runners used:

# Desired Behavior

`molecule test` should successfully perform testing.

# Actual Behaviour

```console
Traceback (most recent call last):
  File "/home/mcdonnnj/.pyenv/versions/ansible-role-docker/bin/molecule", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/molecule/command/test.py", line 113, in test
    base.execute_cmdline_scenarios(scenario_name, args, command_args, ansible_args)
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/molecule/command/base.py", line 119, in execute_cmdline_scenarios
    execute_scenario(scenario)
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/molecule/command/base.py", line 162, in execute_scenario
    execute_subcommand(scenario.config, action)
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/molecule/command/base.py", line 152, in execute_subcommand
    return command(config).execute(args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/molecule/logger.py", line 189, in wrapper
    rt = func(*args, **kwargs)
         ^^^^^^^^^^^^^^^^^^^^^
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/molecule/command/dependency.py", line 40, in execute
    self._config.dependency.execute()
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/molecule/dependency/ansible_galaxy/__init__.py", line 75, in execute
    invoker.execute()
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/molecule/dependency/ansible_galaxy/base.py", line 129, in execute
    self.execute_with_retries()
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/molecule/dependency/base.py", line 55, in execute_with_retries
    util.run_command(self._sh_command, debug=self._config.debug, check=True)
  File "/home/mcdonnnj/.pyenv/versions/3.11.3/envs/ansible-role-docker/lib/python3.11/site-packages/molecule/util.py", line 149, in run_command
    result = app.runtime.exec(
             ^^^^^^^^^^^^^^^^
AttributeError: 'Runtime' object has no attribute 'exec'
```

This is likely caused by  https://github.com/ansible-community/molecule/blob/0d874bd9ccacda8df5e110020454764a6b9b908d/src/molecule/util.py#L149 not being updated for https://github.com/ansible/ansible-compat/pull/256 which was included in [ansible-compat v4.0.1](https://github.com/ansible/ansible-compat/releases/tag/v4.0.1).

### PR DESCRIPTION
**Pull request by asmacdo**
Fixes #3903

https://github.com/ansible/ansible-compat/pull/256

I made this change direct in my site-packages and it fixed the issue i was having. (3903)

Let me know if it needs anything else!

### CODE REVIEW

**Comment by asmacdo** [comment]
Bump to ansible-compat 4.0.1 causes other dependency issues. I guess we could require `ansible-compat<=4.0.0`

I'll wait for a maintainer to suggest what to do next. (I'm happy to do it, but if you'd rather not wait go ahead :)

**Comment by ssbarnea** [comment]
@asmacdo Please continue on this, we do want to switch to the new version and the API changes are not so many, mainly is about switching to `Path` objects for paths and `exec` being renamed to `run`.

For the moment I made a v.5.0.1 hotfix which prevents used of newer compat, at least until we can get your fix in.

**Comment by asmacdo** [comment]
@ssbarnea I think we're good to merge!

**Comment by asmacdo** [comment]
Rebased for conflicts, and fixed the commit messages (I had exec<->run mixed up in text, but correct in code)

**Comment by ziegenberg** [comment]
@asmacdo PR #3924 could partially fix the tox / py311-devel tests.

**Comment by zhan9san** [comment]
@ziegenberg  and @asmacdo 

The CI failed job won't pass until both of your changes are combined in to one.

Besides, the jsondecode issue is fix in [ansible-compat 4.1.2](https://github.com/ansible/ansible-compat/releases/tag/v4.1.2)

For #3924, if `ansible-compat` is directly upgraded to 4.1.2, the `exec<->run` issue will occur.
For #3094, if we want fix CI issue, we need to the change in #3924.

**Comment by zhan9san** [comment]
> Should I bump ansible-compat to 4.1.2 here?

Yes, 4.1.2 is required to fix the json decode issue.

LGTM. Thanks @asmacdo 

@ziegenberg  @ssbarnea 

Do you have any concern?

File: .config/requirements.in
Comment (ziegenberg) [inline .config/requirements.in:1]: We need at least version 4.1.2

**Review by ziegenberg — CHANGES_REQUESTED** [review]

**Comment by asmacdo** [comment]
bumped, recompiled requirements, rebased. Should be good to go if CI is happy :crossed_fingers: 

ty @zhan9san @ziegenberg

**Comment by ziegenberg** [comment]
@asmacdo could you have a look and fix the CI problems?

File: .config/requirements-lock.txt
Comment (ziegenberg) [inline .config/requirements-lock.txt:2]: This change is unnecessary and brakes linting.

File: .config/requirements.txt
Comment (ziegenberg) [inline .config/requirements.txt:2]: This change is unnecessary and brakes linting.

**Review by ziegenberg — COMMENTED** [review]
@asmacdo CI is now almost happy.

File: .config/requirements-lock.txt
Comment (asmacdo) [reply .config/requirements-lock.txt:2]: @ziegenberg That was autogenerated, I'm using 3.10. 

I've got a wip to add docs for how to update deps, tbh I found it a little confusing. Can you verify that this is the correct procedure?


### Updating Dependencies

Dependencies need to be updated by hand in:
  - `.config/requirements.in`
  - `.pre-commit-config.yaml` (2 places)

Afterwards, you will need to generate changes to `requirements.lock.txt`
and `requirement.txt`, by running the commands listed at the top of those files.

Please note that CI will attempt to regenerate those changes, and if there is any diff, CI will fail.

Comment (ziegenberg) [reply .config/requirements-lock.txt]: You should run `tox -e lint` before pushing. That would save you some CI failures.

**Review by ziegenberg — APPROVED** [review]

**Comment by alessfg** [comment]
Any chance you could create a new release to allow us to use the latest versions of Molecule and Ansible Lint simultaneously? 😄

**Comment by hswong3i** [comment]
@alessfg you may manually patch the installed molecule v5.0.1, e.g.
```
$ sudo -E podman run -ti --rm ubuntu:22.04 bash

root@38a5c29f4af4:/# apt update && apt -y full-upgrade && apt -y install python3-dev python3-pip python3-venv patchutils curl

root@38a5c29f4af4:/# pip install pipx

root@38a5c29f4af4:/# pipx install --include-deps molecule

root@38a5c29f4af4:/# cd /root/.local/pipx/venvs/molecule/lib/python3.10/site-packages/molecule/

# We need `patch -p3 -f` here because different folder structure after install by pip
root@38a5c29f4af4:~/.local/pipx/venvs/molecule/lib/python3.10/site-packages/molecule# curl -skL https://github.com/ansible-community/molecule/pull/3904.patch | patch -p3 -f

root@38a5c29f4af4:~/.local/pipx/venvs/molecule/lib/python3.10/site-packages/molecule# fgrep -nH app.runtime.run util.py
util.py:149:    result = app.runtime.run(
```

Well, I had a patched version with .deb/.rpm here, if you could agree with my greedy installation method, too (use with caution, you were warned):
* https://build.opensuse.org/package/show/home:alvistack/ansible-community-molecule-5.0.1
* https://build.opensuse.org/package/show/home:alvistack/ansible-ansible-lint-6.17.1

**Comment by alessfg** [comment]
I could, but it would be even nicer to get a new official release! 😛

**Pull request merged**
